server: grant coordinators call GetMetrics on closed Pebble instances

I attempted to make a change in Pebble which cleared out the `fileCacheHandle` on `Close`

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x1156aba]

goroutine 61812 [running]:
github.com/cockroachdb/pebble.(*fileCacheHandle).metrics(0xc00301eb40)
	external/com_github_cockroachdb_pebble/file_cache.go:192 +0x1a
github.com/cockroachdb/pebble.(*DB).Metrics(0xc00baedb08)
	external/com_github_cockroachdb_pebble/db.go:2051 +0xe2a
github.com/cockroachdb/cockroach/pkg/storage.(*Pebble).GetMetrics(_)
	pkg/storage/pebble.go:1842 +0x65
github.com/cockroachdb/cockroach/pkg/server.(*nodePebbleMetricsProvider).GetPebbleMetrics.func1(0xc004b21808)
	pkg/server/node.go:1391 +0x74
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).VisitStores.func1(0x8f6d150?, 0xc0061e6150?)
	pkg/kv/kvserver/stores.go:146 +0x26
github.com/cockroachdb/cockroach/pkg/util/syncutil.(*Map[...]).Range(0x8bd1ce0?, 0xc00b223d38)
	pkg/util/syncutil/map.go:391 +0x10d
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).VisitStores(0x49f46e?, 0x1b3e47d?)
	pkg/kv/kvserver/stores.go:145 +0x55
github.com/cockroachdb/cockroach/pkg/server.(*nodePebbleMetricsProvider).GetPebbleMetrics(0xc00a41cf48)
	pkg/server/node.go:1389 +0x1a8
github.com/cockroachdb/cockroach/pkg/util/admission.(*StoreGrantCoordinators).SetPebbleMetricsProvider.func1()
	pkg/util/admission/grant_coordinator.go:120 +0x1bb
created by github.com/cockroachdb/cockroach/pkg/util/admission.(*StoreGrantCoordinators).SetPebbleMetricsProvider in goroutine 59280
	pkg/util/admission/grant_coordinator.go:96 +0x389
```


The engine closer is registered here:
https://github.com/cockroachdb/cockroach/blob/cada6b4976accfefa486b9e81804b2193f320fe8/pkg/server/server.go#L298

The grant coordinator closer is here:
https://github.com/cockroachdb/cockroach/blob/cada6b4976accfefa486b9e81804b2193f320fe8/pkg/server/server.go#L599

There are two problems:
 - closers don't run in reverse order (which seems wrong in general)
 - the grant coordinator closer doesn't actually wait for the background loop to exit (https://github.com/cockroachdb/cockroach/blob/cc8f6678e08b217c5926b6e8e82da774168c1739/pkg/util/admission/grant_coordinator.go#L278)

Jira issue: CRDB-47168

--- a/pkg/util/admission/grant_coordinator.go
+++ b/pkg/util/admission/grant_coordinator.go
@@ -276,6 +276,14 @@ func (sgc *StoreGrantCoordinators) TryGetSnapshotQueueForStore(storeID roachpb.S
 func (sgc *StoreGrantCoordinators) close() {
 	// closeCh can be nil in tests that never called SetPebbleMetricsProvider.
 	if sgc.closeCh != nil {
+		// Ensure that the goroutine has observed the close and will no longer
+		// call GetPebbleMetrics, since the engines will be closed soon after this
+		// method returns, and calling GetPebbleMetrics on closed engines is not
+		// permitted.
+		sgc.closeCh <- struct{}{}
+		// Close the channel, so that if close gets called twice due to a bug,
+		// sending on the closed channel will panic instead of the send being
+		// blocked forever.
 		close(sgc.closeCh)
 	}
 

